### PR TITLE
Add json output format for crc status

### PR DESCRIPTION
**Pull request by guillaumerose**
Partial fix for #1227

```
$ crc status -o json               
{
  "virtualMachineStatus": "Running",
  "openshiftStatus": "Running",
  "openshiftVersion": "4.5.1",
  "diskUsage": 12852875264,
  "diskSize": 32720793600,
  "cacheUsage": 9769732330,
  "cacheDir": "/home/guillaumerose/.crc/cache"
}

$ crc status        
CRC VM:          Running
OpenShift:       Running (v4.5.1)
Disk Usage:      12.86GB of 32.72GB (Inside the CRC VM)
Cache Usage:     9.77GB
Cache Directory: /home/guillaumerose/.crc/cache

$ crc status -o wide
CRC VM:          Running
OpenShift:       Running (v4.5.1)
Disk Usage:      12.86GB of 32.72GB (Inside the CRC VM)
Cache Usage:     9.77GB
Cache Directory: /home/guillaumerose/.crc/cache

```

### CODE REVIEW

File: cmd/crc/cmd/status.go [hunk 1]
```diff
@@ -34,23 +45,24 @@ Cache Directory: {{.CacheDir}}
 `
 
 type Status struct {
-	CrcStatus       string
-	OpenShiftStatus string
-	DiskUsage       string
-	CacheUsage      string
-	CacheDir        string
+	CrcStatus       string `json:"crcStatus"`
```
Comment (gbraad) [inline cmd/crc/cmd/status.go:48]: We do not care much about the actual markup, but rather about providing the needed information. This would be more important is we would ever rename something.

File: cmd/crc/cmd/status.go [hunk 2]
```diff
@@ -59,23 +71,30 @@ func runStatus() {
...
+		return fmt.Errorf("Error finding size of cache: %s", err.Error())
 	}
 	cacheUsage := units.HumanSize(float64(size))
 	diskUse := units.HumanSize(float64(clusterStatus.DiskUse))
 	diskSize := units.HumanSize(float64(clusterStatus.DiskSize))
 	diskUsage := fmt.Sprintf("%s of %s (Inside the CRC VM)", diskUse, diskSize)
 	status := Status{clusterStatus.CrcStatus, clusterStatus.OpenshiftStatus, diskUsage, cacheUsage, cacheDir}
-	printStatus(status, statusFormat)
+
+	if outputFormat == jsonFormat {
```
Comment (gbraad) [inline cmd/crc/cmd/status.go:82]: this is more generic, so the writer is better suited to be created in a generic way as ALL commands will use this.

Comment (guillaumerose) [reply cmd/crc/cmd/status.go:81]: I prefer to start small and don't over-engineer it in this PR. But for sure, when we will have 2 commands with `-o json` we will need to do something.

File: cmd/crc/cmd/status.go (hunk 1 above)
Comment (guillaumerose) [reply cmd/crc/cmd/status.go]: Markup is better if it's consistent with other products. OpenShift and Kubernetes use lower camel case. CRC should do the same.

File: cmd/crc/cmd/status.go [hunk 3]
```diff
@@ -90,17 +83,31 @@ func runStatus(writer io.Writer, client client, cacheDir, outputFormat string) e
...
+func printStatus(writer io.Writer, status Status) error {
+	w := tabwriter.NewWriter(writer, 0, 0, 1, ' ', 0)
+	lines := []struct {
+		left, right string
+	}{
+		{"CRC VM", status.CrcStatus},
+		{"OpenShift", status.OpenShiftStatus},
+		{"Disk Usage", status.DiskUsage},
+		{"Cache Usage", status.CacheUsage},
+		{"Cache Directory", status.CacheDir},
```
Comment (praveenkumar) [inline cmd/crc/cmd/status.go:102]: I would also like some more details about this change, like what is the benefit of using tabwritter over templates.

File: cmd/crc/cmd/status.go [hunk 4]
```diff
@@ -38,10 +38,11 @@ var statusCmd = &cobra.Command{
 }
 
 type Status struct {
-	CrcStatus       string `json:"crcStatus"`
+	CrcStatus       string `json:"virtualMachineStatus"`
```
Comment (praveenkumar) [inline cmd/crc/cmd/status.go]: This change don't belong to this commit.

File: cmd/crc/cmd/status.go [hunk 5]
```diff
@@ -68,12 +69,13 @@ func runStatus(writer io.Writer, client client, cacheDir, outputFormat string) e
...
-		DiskSize:        clusterStatus.DiskSize,
-		CacheUsage:      size,
-		CacheDir:        cacheDir,
+		CrcStatus:        clusterStatus.CrcStatus,
+		OpenShiftStatus:  clusterStatus.OpenshiftStatus,
+		OpenShiftVersion: clusterStatus.OpenshiftVersion,
+		DiskUsage:        clusterStatus.DiskUse,
+		DiskSize:         clusterStatus.DiskSize,
+		CacheUsage:       size,
+		CacheDir:         cacheDir,
```
Comment (praveenkumar) [inline cmd/crc/cmd/status.go:78]: This is again public facing UX change so I would like to have discussion about it before merge this one.

File: cmd/crc/cmd/status.go (hunk 3 above)
Comment (guillaumerose) [reply cmd/crc/cmd/status.go:102]: In the following commit, I change the struct to remove strings and have raw numbers for usage. It was easier to build the original template by hand than with template.

File: cmd/crc/cmd/status.go (hunk 5 above)
Comment (guillaumerose) [reply cmd/crc/cmd/status.go:78]: The status message didn't change. With json output, for sure, now we cannot change it like this.

But for now, I don't think someone is relying on this struct ? Does someone vendor crc ?

File: cmd/crc/cmd/status.go (hunk 4 above)
Comment (guillaumerose) [reply cmd/crc/cmd/status.go]: yes you're right. I was not sure of this change. WDYT of virtualMachineStatus ?

File: cmd/crc/cmd/status.go (hunk 5 above)
Comment (praveenkumar) [reply cmd/crc/cmd/status.go:78]: > Does someone vendor crc ?

@guillaumerose not that I am aware of.

File: cmd/crc/cmd/status.go (hunk 3 above)
Comment (praveenkumar) [reply cmd/crc/cmd/status.go:102]: @guillaumerose Can you put that in the commit log the benefit it is going to provide moving forward or right now.

Comment (guillaumerose) [reply cmd/crc/cmd/status.go:102]: I added the why in the commit log

File: cmd/crc/cmd/status.go (hunk 5 above)
Comment (anjannath) [reply cmd/crc/cmd/status.go:78]: The trays depend on the struct, so will need changes there

Comment (guillaumerose) [reply cmd/crc/cmd/status.go:78]: Yes, looking at the macOS tray we use CrcStatus, OpenshiftStatus, DiskUse and DiskSize :
```
➜  tray-macos git:(master) rg "status\."
CodeReady Containers/DetailedStatusViewController.swift
44:                    if status.Success {
45:                        self.vmStatus.stringValue = status.CrcStatus
46:                        self.ocpStatus.stringValue = status.OpenshiftStatus
47:                        self.diskUsage.stringValue = "\(Units(bytes: status.DiskUse).getReadableUnit()) of \(Units(bytes: status.DiskSize).getReadableUnit()) (Inside the VM)"
```

This PR (apart from breaking the json due the case) removes the version from OpenshiftStatus. 

What should we do ? The tray is still experimental so we can edit in both and release at the same time. WDYT of the case of the json ? I took the same convention as kubernetes/openshift.

Comment (guillaumerose) [reply cmd/crc/cmd/status.go:78]: In fact this struct is not used by the tray. This struct is only use in cmd package. Handler use ClusterStatusResult. So we are good.

File: cmd/crc/cmd/client.go [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+type client interface {
+	Status(statusConfig machine.ClusterStatusConfig) (machine.ClusterStatusResult, error)
+	Exists(name string) (bool, error)
+}
+
+type libmachineClient struct{}
+
+func (*libmachineClient) Status(statusConfig machine.ClusterStatusConfig) (machine.ClusterStatusResult, error) {
+	return machine.Status(statusConfig)
+}
```
Comment (praveenkumar) [inline cmd/crc/cmd/client.go:14]: @guillaumerose this new `client` for the cmd is only for unit test purpose so that it can be mocked and also extended for other commands ( like start,stop ..etc.) ?

Comment (guillaumerose) [reply cmd/crc/cmd/client.go:14]: Yes !

File: cmd/crc/cmd/status.go (hunk 5 above)
Comment (anjannath) [reply cmd/crc/cmd/status.go:78]: Yeah, sorry.. i thought i commented on the `ClusterStatusResult` structure changes, we can change in both, that's no issue. and i am fine with the case of the json, although i find Upper camel case more good to look at..

File: cmd/crc/cmd/status.go (hunk 4 above)
Comment (anjannath) [reply cmd/crc/cmd/status.go]: i'd prefer it to be called `crcStatus` only, as this also relates to the struct member name, if that makes sense..

File: cmd/crc/cmd/status.go [hunk 7]
```diff
@@ -14,7 +15,14 @@ import (
...
 )
 
+const jsonFormat = "json"
+
+var (
+	outputFormat string
+)
+
 func init() {
+	statusCmd.Flags().StringVarP(&outputFormat, "output", "o", "", "Output format. One of: json|wide")
```
Comment (anjannath) [inline cmd/crc/cmd/status.go:25]: The `wide` format can be removed, as its not been implemented.

**Comment by anjannath** [comment]
Looks good to me 👍

File: cmd/crc/cmd/status.go (hunk 7 above)
Comment (guillaumerose) [reply cmd/crc/cmd/status.go:25]: Yes good point. What could be the name of the default format ? For kubectl, it's wide if I'm correct.

File: cmd/crc/cmd/status.go (hunk 4 above)
Comment (guillaumerose) [reply cmd/crc/cmd/status.go]: I put back `crcStatus`

File: cmd/crc/cmd/status.go (hunk 7 above)
Comment (anjannath) [reply cmd/crc/cmd/status.go:25]: `kubectl` wide gives more information, then without a `-o` flag which is the default. i don't think we need to provide the default also as option for the output flag.

Comment (guillaumerose) [reply cmd/crc/cmd/status.go]: I removed wide and just let json. Is it better? `Output format. One of: json` ?
